workers/unity/Assets/Gdk/Core/Utility/ConnectionUtility.cs

```diff
@@ -89,45 +75,67 @@ private static ConnectionParameters CreateConnectionParameters(ConnectionConfig
             return connectionParameters;
         }
 
-        private static Connection TryToConnect(Future<Connection> connectionFuture)
+        private static Connection TryToConnect(Future<Connection> connectionFuture, out ConnectionError error)
```

Comment:
This feels odd to me.
Why not return a struct something like
```
struct ConnectionResult
{
	private Connection connection;
	public Error? {get;}
    public 	Connection {get {Error.HasValue ? throw : connection }  }
}
```

Reply:
One issue with this approach which I have is that I also call `string GetDeploymentName(Locator locator, out ConnectionError error)` within `TryToConnect`. So I would also have to change this to return a struct with a string and an error to be consistent. I think thats uglier.

Reply:
I strongly disagree. Using an `out` parameter to return an error isn't idiomatic C#. Throwing an exception is, but apparently we don't want to do that for some reason.

Reply:
@samiwh disagreed with throwing an exception when it is expected that it can happen and should be part of the normal workflow.

Reply:
I agree with that in general. The exact use case for the error context object was, if I remember correctly, when an error happens in one of the systems. In that case throwing an exception might not work as expected, but you'll want to handle that error in some way.

Here: I agree, would prefer to throw an exception

Reply:
If we think errors codes will be harder to write or more confusing to users because they are less idiomatic then I'm happy to use exceptions instead. Exceptions also play very slightly nicer with tasks (and we have played around with making connect an async task).

There are two main reasons I was against using exceptions.

1. These failures aren't exceptional, they are just normal control flow; connections can sometimes fail when everything is configured correctly. With a file that usually means it needs user intervention to correct. I'd argue this is more akin to `Dictionary.TryGetValue`.
If we disagree on this then falling back to exceptions seems like a reasonable default.

2. I personally find try catch blocks hard to read, and retry logic for things that throw exceptions can look odd.
(However I will conceded that the code for generating the error often looks better)

Reply:
The other reason was unity catching all exceptions and making them easy to miss.

Reply:
If you want to see the possible implementation of using exceptions its in #c68689a. If we decide to go with error objects we can reverse it.